contracts/Auction.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity >=0.5.0 < 0.9.0;
+
+contract Auction{
+    // who started auction, and will get paid
+    address payable public owner;
+    uint public startBlock;
+    uint public endBlock;
+
+    //off-chain info about item -- on IPFS
+    //keep pointer to IPFS hash
+    string public ipfsHash;
+
+    enum State {Started, Running, Ended, Canceled}
+    State public auctionState;
+
+    uint public highestBindingBid;
+    address payable public highestBidder;
+
+    // keeping addresses, amounts of bidders in a map
+    mapping (address => uint) public bids;
+
+    // bit increments
+    uint bidIncrement;
+
+
+    // auction is owneed by whomever initiates the contract
+    constructor() {
+        owner = payable(msg.sender);
+        auctionState = State.Running;
+
+        // AUCTION DURATION -- declare how many blocks to allow until auction stops
+        // 24 hr auction, so need current block plus a day's worth of blocks
+        // a block created about every 15 sec, so 86400 / 15 sec is a day
+        startBlock = block.number;
+        endBlock = startBlock + 5760;
+        ipfsHash = "";
+
+        // bid increment in wei
+        bidIncrement = 100;
+    }
+
+
+    // modifiers to enforce some business rules
+    modifier onlyOwner(){
+        require(msg.sender == owner);
+        _;
+    }
+    modifier notOwner(){
+        require(msg.sender != owner);
+        _;
+    }
+
+    modifier afterStart(){
+        require(block.number >= startBlock);
+        _;
+    }
+
+    modifier beforeEnd(){
+        require(block.number <= endBlock);
+        _;
+    }
+
+    // figure out the min of two bids
+    function min(uint x, uint y) pure internal returns (uint){
+        if (x < y){
+            return x;
+        }
+        else {
+            return y;
+        }
+    }
+
+
+    function placeBid() public payable notOwner afterStart beforeEnd {
+        require(auctionState == State.Running);
+        require(msg.value >= bidIncrement);
+
+        // calculate currentBid, ensure its big enough
+        uint currentBid = bids[msg.sender] + msg.value;
+        require(currentBid > highestBindingBid);
+
+        bids[msg.sender] = currentBid;
+
+        // is this bid now highest biding, and greater than current highest bidder?
+        if(currentBid <= bids[highestBidder]){
+            highestBindingBid = min(currentBid + bidIncrement, bids[highestBidder]);
+        } else {
+            highestBindingBid = min(currentBid, bids[highestBidder] + bidIncrement);
+            highestBidder = payable(msg.sender);
+        }
```

Comment:
generally, you want to issue an event so off-chain indexing tools can update on events, if no event is declared you have no way of knowing unless you're polling a public variable.

Reply:
same comments for all functions

Reply:
I follow what you're saying as it makes sense. Do you have any concrete examples to refer to? I'm used to data-flow, or event-streams with traditional api backend architecture, but i'm not sure how to map that to this evm/contract/dapp space yet.

Reply:
https://www.tutorialspoint.com/solidity/solidity_events.htm

Reply:
https://thegraph.com/docs/en/about/introduction/ This explains one way to pull the event data off-chain for indexing so the web2 side can utilize the information

Reply:
Cool... taking a look.. thanks!